### ISSUE TITLE
Data Provider callbacks should include the xpath that user has requested

### ISSUE THREAD

**Issue by bharath154** [sysrepo/sysrepo#885]
### Things you may try first
(put "x" in "[]" if you already tried following)
* [x] Did you check if this is a duplicate issue?
* [x] Did you test it on the latest sysrepo devel branch?

### Description
When we subscribe to the xpath using "sr_dp_get_items_subscribe" I understand that the limitation is we cannot subscribe to an xpath which contains a list with keys, for example /ietf-interfaces/ietf-interface[name='UNI-1'].

But because of this limitation, there can be huge performance problems since we are forced to fill in all the details of the sub-tree, since we will end up receiving call backs for every container in that sub-tree.

Now, my questions/proposals are 
1) Why can't we pass the target xpath that user has requested to this function, so that user can create the relevant sub-tree rather than every possible container under /ietf-interfaces i.e., have to create for all the interfaces present in the system like UNI-1, UNI-2, UNI-3, UNI-4 etc.,

2) Is it possible to read a user defined xml file for read-only (operational) data and create the complete tree and may be then we can subscribe to particular xpath rather than creating this tree dynamically every time?

**Comment by michalvasko** [comment]
Hi,

1. What I can imagine is having another function for subscribing to provide module data, which would be forced to always provide data of the whole module. It would then be possible to pass `get`/`get-config` filter to the callback registered this way and it would be up to the application to parse it and return only what is needed. However, this would allow buggy applications to return more/less data that is correct, so I am not really sure it is a good idea. You see, your problem is relevant, but the solution must be a good one. Currently it may not work most efficiently, but it works correctly.

I can also imagine some sort of caching system when the application first provides data and at later request it would be given the option to say that the previous data are still valid. Anyway, this all would be an enhancement and we will not work on them before all the features are finished. You are welcome to provide patches, though.

2. I am not sure I understand you proposal, you want to create the state data once, store them in a file, and then have *sysrepo* read the file whenever this data is requested? And you would update the file every time the state data changes? I see only problems with this solution. Mainly, you are basically substituting GPB communication for a file, which would hardly be more effective. Also, not an uncommon use-case of having some counters in state data would be extremely inefficient having to constantly update the file.

To summarize, I do think this is an issue, but either implement a robust solution yourself or leave it up to us even though it will not get implemented any time soon.

Regards,
Michal

**Comment by jjin62** [comment]
Hi Michal,
I have related observation here. For example, in oper_data_example.c:

seems regardless what the xpath specified in the data_requester(), sysrepo always call the data_provider() multiple times with the same set of xpaths.

in data_request():
1) rc = sr_get_items_iter(session, "/ietf-interfaces:interfaces-state/interface//*", &iter);
2) rc = sr_get_items_iter(session, "/ietf-interfaces:interfaces-state/interface[name='eth0']", &iter);
3) rc = sr_get_items_iter(session, "/ietf-interfaces:interfaces-state/interface[name='eth0']/oper-status", &iter);

The result are correct:
1) $ ./oper_data_example  1
Requesting state data of ietf-inetrfaces:
/ietf-interfaces:interfaces-state/interface[name='eth0']/name = eth0
/ietf-interfaces:interfaces-state/interface[name='eth0']/oper-status = down
/ietf-interfaces:interfaces-state/interface[name='eth0']/statistics (container)
/ietf-interfaces:interfaces-state/interface[name='eth0']/statistics/discontinuity-time = 1987-08-31T17:00:30.44Z
/ietf-interfaces:interfaces-state/interface[name='eth1']/name = eth1
/ietf-interfaces:interfaces-state/interface[name='eth1']/type = iana-if-type:ethernetCsmacd
/ietf-interfaces:interfaces-state/interface[name='eth1']/oper-status = up
/ietf-interfaces:interfaces-state/interface[name='eth1']/statistics (container)
/ietf-interfaces:interfaces-state/interface[name='eth1']/statistics/discontinuity-time = 2016-10-06T15:12:50.52Z

2) $ ./oper_data_example  1
Requesting state data of ietf-inetrfaces:
/ietf-interfaces:interfaces-state/interface[name='eth0']/name = eth0
/ietf-interfaces:interfaces-state/interface[name='eth0']/oper-status = down
/ietf-interfaces:interfaces-state/interface[name='eth0']/statistics (container)
/ietf-interfaces:interfaces-state/interface[name='eth0']/statistics/discontinuity-time = 1987-08-31T17:00:30.44Z

3) $ ./oper_data_example  1
Requesting state data of ietf-inetrfaces:
/ietf-interfaces:interfaces-state/interface[name='eth0']/oper-status = down

**But data_provider() are called in the same way for above 3 cases:**
$ ./oper_data_example 
This application will be a data provider for state data of ietf-interfaces.
Run the same executable with one (any) argument to request some data.


 ========== SUBSCRIBED FOR PROVIDING OPER DATA ==========


Data for '/ietf-interfaces:interfaces-state' requested.
Data for '/ietf-interfaces:interfaces-state/interface' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/statistics' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/statistics' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/ietf-ip:ipv4' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/ietf-ip:ipv4' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/ietf-ip:ipv6' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/ietf-ip:ipv6' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/ietf-ip:ipv4/address' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/ietf-ip:ipv4/neighbor' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/ietf-ip:ipv4/address' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/ietf-ip:ipv4/neighbor' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/ietf-ip:ipv6/address' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth0']/ietf-ip:ipv6/neighbor' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/ietf-ip:ipv6/address' requested.
Data for '/ietf-interfaces:interfaces-state/interface[name='eth1']/ietf-ip:ipv6/neighbor' requested.

Is it possible for sysrepo to invoke callbacks more intelligently based on xpath in data_request? or just pass the exact xpath requested by data_requester() to data_provider()?

**In our case, we have a lager list (>100 elements)**, So every time the data provider has to return the whole list (500k data), even the client only request one entry.

thanks,
Jimmy

**Comment by michalvasko** [comment]
Not relevant/fixed.
